### PR TITLE
Add support for video profiles

### PR DESCRIPTION
**Pull request by jehumb-havok**
Add the ability to select the video profile of a video capture device by unique string identifier, and to select a particular resolution and/or framerate within that profile.

For consistency, also add the ability to select the resolution and/or framerate when not using video profiles, as it seems that WebRTC media constraints are deprecated and being removed : https://bugs.chromium.org/p/webrtc/issues/detail?id=9239

### CODE REVIEW

**Comment by jehumb-havok** [comment]
Ping @tadam-msft - you're the only alias I know, sorry.

**Comment by tadam-msft** [comment]
Sorry in training all day tidy, I'll try to take a look.

**Review by robin-raymond — APPROVED** [review]
Obviously this needs to be tested thoroughly first but the code changes are good from a visual perspective. @morosev can you review too please since this is most where you've been coding?

**Comment by jehumb-havok** [comment]
Updated the code to use [GetAvailableMediaStreamProperties()](https://docs.microsoft.com/en-us/uwp/api/windows.media.devices.videodevicecontroller.getavailablemediastreamproperties) for supported video formats, and enforce consistency between WebRTC and MediaCapture.

First, when a MediaCapture object is initialized with a video profile, then GetAvailableMediaStreamProperties() will filter out the media types (= video formats) returned to those compatible with the selected profile. This makes the code much simpler and naturally consistent. This is also the recommended way according to the Media Foundation team.

However, WebRTC and MediaCapture have different ways of selecting a media type / video format to use:
- For MediaCapture, setting the [RecordMediaDescription](https://docs.microsoft.com/en-us/uwp/api/windows.media.capture.mediacaptureinitializationsettings.recordmediadescription) field on the init settings instructs the driver to use that media type by default. But WebRTC doesn't know about this.
- For WebRTC, the video format is selected with a heuristic involving the list of supported video formats returned by the device, and a set of media constraints.

There are 2 issues with this:
1. WebRTC and MediaCapture are not aware of each other's choice, and it can happen that MediaCapture opens the video device with a different resolution than the one expected by WebRTC, which will result in video corruption or crashes.
2. The media constraints API is deprecated, and removed in later versions of WebRTC, so is not the way to go.

As a result, the proposed change will only return to WebRTC a single supported video format corresponding to the media type selected in MediaCapture, which locks WebRTC into the only good choice of video format and prevents any possible discrepancy.

**Review by robin-raymond — COMMENTED** [review]
Will do a final review when changes are ready.

**Review by morosev — APPROVED** [review]
Everything looks good to me. The selection of video properties is definitely better than it was before. We might use the same engine in M75, since the video deice format selection has been changed there as well.

**Comment by jehumb-havok** [comment]
Added support for [`KnownVideoProfile`](https://docs.microsoft.com/en-us/uwp/api/windows.media.capture.knownvideoprofile) which is nicer than GUID-based profile IDs when the app displays some UI, and is also device-independent.

Tested on HoloLens 2 (with profiles) and locally on PC with a USB webcam (no profile). This is ready for a final review @robin-raymond. Ping also @tadam-msft if you want to have a look.

**Comment by jehumb-havok** [comment]
Ping @robin-raymond and @tadam-msft, can you please have a look before we merge? We have customers waiting for those changes.

**Comment by robin-raymond** [comment]
@djee-ms Ok, I looked preliminary a while ago but I'll relook again today. Sorry I was away for a few days on holiday.

File: idl/VideoCapturer.idl [hunk 1]
```diff
@@ -93,13 +93,66 @@ namespace org
       int height;
     };
 
+	enum VideoProfileKind
```
Comment (morosev) [inline idl/VideoCapturer.idl]: Tab is used instead of spaces

File: windows/wrapper/impl_webrtc_IVideoCapturer.h [hunk 2]
```diff
@@ -61,7 +61,12 @@ namespace webrtc
 
       const char *name_ {};
       const char *id_ {};
+      const char *videoProfileId_ {};
+	    int videoProfileKind_ {};
```
Comment (morosev) [inline windows/wrapper/impl_webrtc_IVideoCapturer.h]: Tab symbol

File: windows/wrapper/impl_webrtc_VideoCapturer.cpp [hunk 3]
```diff
@@ -838,91 +973,257 @@ namespace webrtc
   VideoCapturerUniPtr VideoCapturer::create(const CreationProperties &info) noexcept
   {
     auto result = std::make_unique<VideoCapturer>(make_private{});
-    result->init(info);
-    return result;
+    if (result->init(info)) {
+	  return result;
+    }
+	return {};
```
Comment (morosev) [inline windows/wrapper/impl_webrtc_VideoCapturer.cpp]: Tab symbol

File: windows/wrapper/impl_webrtc_VideoCapturer.cpp [hunk 4]
```diff
@@ -838,91 +973,257 @@ namespace webrtc
...
     }
+    if (formats.empty()) {
+      RTC_LOG_F(LS_ERROR)
+          << "Failed to find any supported video format for video capture device "
+          << device_unique_id_utf8
+          << ". Try relaxing constraints on video profile, resolution, or "
+             "framerate";
+      device_ = nullptr;
+      return false;
+	}
```
Comment (morosev) [inline windows/wrapper/impl_webrtc_VideoCapturer.cpp]: Tab symbol

File: windows/wrapper/impl_webrtc_VideoCapturer.cpp [hunk 5]
```diff
@@ -838,91 +973,257 @@ namespace webrtc
...
     SetSupportedFormats(formats);
+	return true;
```
Comment (morosev) [inline windows/wrapper/impl_webrtc_VideoCapturer.cpp]: Tab symbol

**Review by morosev — APPROVED** [review]
I put just few comments related to formatting - using tab symbol instead of spaces. All other looks good to me.

**Comment by robin-raymond** [comment]
The changes overall look good. I think this is likely to break dot net implementation. I'd have to clone it down, merge in, and compile the dot net to see what happens.

**Comment by jehumb-havok** [comment]
@robin-raymond why do you think that? Video profiles are optional as implemented, so existing code should work. We are discussing with @jamescadd going ahead with the merge due to time constraints, so if .NET imlementation is going to break this will be a blocker.

**Comment by robin-raymond** [comment]
Go ahead and merge. I will fix releases/m71 compilation issues should any occur. I don't expect any runtime issue at all.

**Pull request merged**